### PR TITLE
fix(streaming): strip orphaned tool_calls from CLI bridge message reconstruction (#3583)

### LINKED ISSUE #3583: 400 Bad Request from upstream API due to orphaned tool_calls in CLI bridge message reconstruction
## Symptom

When using Web UI (CLI bridge mode / `dispatch_in_gateway: true`), the browser console occasionally shows a 400 error:

```
POST http://127.0.0.1:8649/api/chat/completions 400 (Bad Request)
```

```json
{
  "error": {
    "message": "The messages array must have an even number of turn pairs. There is one assistant message with 'tool_calls' that has no matching 'tool' message(s) after it. tool_call_id: call_XXXXX",
    "type": "invalid_request_error"
  }
}
```

The error comes directly from the upstream API (DeepSeek/OpenAI). The Web UI passes it through to the browser.

## Root Cause

The CLI bridge has a message sync issue between the Hermes CLI's in-memory session and the Web UI's message reconstruction. Key evidence from server logs:

### 1. Memory/DB message count mismatch
```
[chat-run-socket] syncFromHermes: memory has 616 messages, DB has 41 messages
[chat-run-socket] syncFromHermes: synced 41 messages to local session
```
The CLI side has 616 messages but only 41 survive the sync to the Web UI DB. Many tool result messages are lost during sync.

### 2. Abort timeout drops tool results
```
[chat-run-socket][abort] CLI bridge interrupt did not sync before timeout
```
When a user aborts a running session, the interrupt signal doesn't wait for the bridge to finish flushing pending tool results. This leaves orphaned `tool_calls` in the message array.

### 3. How it manifests
When the Web UI reconstructs the messages array to send to the upstream API, some `tool` role messages (tool results) are missing. The API sees an assistant message with `tool_calls` but no corresponding `tool` message with a matching `tool_call_id`, and returns 400.

## Environment

- Web UI version: 0.6.10
- Profile: default, `dispatch_in_gateway: true`
- Bridge: CLI bridge (`hermes_bridge.py` via IPC)
- Upstream: DeepSeek API (any API with strict message turn-pair validation)
- Setup: WSL + Windows

## Crash Course Reproduction

The error is intermittent, but more likely to trigger when:
1. The model makes many consecutive tool calls (terminal, patch, search_files, etc.)
2. The user aborts mid-run and immediately sends a new message
3. The bridge is under heavy load

## Suggested Fixes

1. **Fix the sync mechanism**: Ensure all pending tool results are flushed from the CLI bridge to the Web UI before the run completes or the session is aborted. The `interrupt did not sync before timeout` log entry suggests the abort sequence doesn't wait for pending tool results.

2. **Add a repair step in the proxy**: Before forwarding messages to the upstream API, strip orphaned `tool_calls` entries that have no corresponding `tool` message. This is a defensive measure for cases where sync still fails.

3. **Increase bridge sync timeout**: The current abort timeout may be too short for sessions with many tool calls.

## Diagnostic Data Available

- `server.log` at `~/.hermes-web-ui/logs/server.log` shows `syncFromHermes` message count gaps
- `bridge.log` at `~/.hermes-web-ui/logs/bridge.log` shows IPC communication
- The 400 error does NOT appear in server.log as a logged error — it's the upstream API response passed to the browser via the `/api/chat/completions` proxy endpoint

### PR DESCRIPTION
**Pull request by rodboev**
Closes #3583

## Thinking Path

- CLI bridge mode aborts can interrupt the bridge before tool results are written to the session store, leaving assistant messages with `tool_calls` entries that have no matching `tool` role responses.
- `_sanitize_messages_for_api` already strips orphaned `tool` responses (dangling results without a matching assistant `tool_calls`), but does not handle the inverse direction.
- Strict providers (DeepSeek, OpenAI) reject the resulting message array with a 400 Bad Request because they require every `tool_calls` entry to have a downstream `tool` response.
- Adding a reverse-direction strip pass after the existing logic closes the gap without changing abort timing or bridge behavior.

## What Changed

- `api/streaming.py`: extended `_sanitize_messages_for_api` and `_api_safe_message_positions` with a third pass that collects answered `tool_call_id`s from surviving `tool` role messages, then strips unanswered entries from assistant `tool_calls` arrays. Messages left with an empty `tool_calls` list have the key removed; content-less messages with no remaining `tool_calls` are dropped entirely.
- `tests/test_issue3583_orphaned_tool_calls.py`: new test file covering basic orphan strip, all-orphaned, no-orphans, mixed, and Anthropic `call_id` format scenarios.

## Why It Matters

Any CLI bridge abort that races tool result persistence breaks subsequent turns for the session on strict providers. This is a silent, session-corrupting failure that requires manual message editing or session abandonment to recover from.

## Verification

```
pytest tests/test_issue3583_orphaned_tool_calls.py -v --timeout=60
pytest tests/ -v --timeout=60
```

## Risks / Follow-ups

- The strip is lossy; orphaned tool calls are silently removed rather than surfaced to the user. A follow-up could log a warning or inject a system note so the user knows context was dropped.
- This does not prevent the root cause (abort timing). A complementary fix in the bridge abort path to flush pending tool results before disconnecting would reduce the frequency of orphaned calls.

## Model Used

Claude Opus 4.8 via Claude Code CLI

### CODE REVIEW

File: api/streaming.py [hunk 1]
```diff
@@ -2684,7 +2684,35 @@ def _sanitize_messages_for_api(messages, *, cfg: dict = None):
...
+            ]
+            if not kept:
+                # All calls orphaned: drop tool_calls key; if no content, drop message.
+                msg = {k: v for k, v in msg.items() if k != 'tool_calls'}
+                if not str(msg.get('content') or '').strip():
+                    continue
+            else:
+                msg = dict(msg, tool_calls=kept)
+        filtered_clean.append(msg)
+    return filtered_clean
```
[the remark this answers is not in the document]
Comment (rodboev) [reply api/streaming.py:2715]: The two functions have parallel structure by design (they mirror each other for different return types). A shared helper would need to abstract over list-of-dicts vs list-of-tuples, adding complexity for two call sites. Noted as a follow-up if a fourth sanitization pass is ever added.

**Comment by nesquena-hermes** [comment]
Reviewed the diff against `api/streaming.py` on `origin/master` plus the consuming call sites and the agent-side sanitizer. The code change itself is correct and safe — but there's a layering question worth raising before merge, because the fix may not sit on the path issue #3583 actually fails on.

## What the diff does (correct, mirrors existing logic)

The third pass in `_sanitize_messages_for_api` collects answered `tool_call_id`s, then strips assistant `tool_calls` whose id has no matching `tool` response — the exact inverse of the existing orphaned-*result* strip. The `id`/`call_id` dual-key handling matches the first pass on master (`api/streaming.py:2650`), and dropping the `tool_calls` key (then the whole message when content is empty) is the right shape:

```python
if not kept:
    msg = {k: v for k, v in msg.items() if k != 'tool_calls'}
    if not str(msg.get('content') or '').strip():
        continue
```

## Layering concern — does this cover the failing path?

Issue #3583's repro is `dispatch_in_gateway: true` (gateway/CLI-bridge mode), and the 400 comes from the upstream `/v1/chat/completions`. But the WebUI gateway path doesn't send reconstructed history through `_sanitize_messages_for_api`. In `api/gateway_chat.py:308` the outbound payload is:

```python
body = {
    "model": model or "default",
    "stream": True,
    "messages": [*prefill_messages, {"role": "user", "content": message_content}],
}
```

Only prefill + the current user message — history continuity is server-side via `X-Hermes-Session-Key`. So orphaned `tool_calls` are reconstructed *gateway-side*, not in the WebUI, and this patch won't be on that wire. Where the patch *does* help is the **direct (non-gateway) agent path**: `api/streaming.py:5529` and `api/routes.py:11309`, both of which call `conversation_history=_sanitize_messages_for_api(_previous_context_messages, ...)`.

## Lossy strip vs. agent's lossless stub-injection

The agent already solves this exact inverse-orphan case, but by a different strategy. `sanitize_api_messages` (`agent/agent_runtime_helpers.py:1842`, invoked unconditionally at `agent/conversation_loop.py:1035`) **injects a stub result** rather than stripping the call:

```python
missing_results = surviving_call_ids - result_call_ids
...
patched.append({
    "role": "tool",
    "content": "[Result unavailable — see context summary above]",
    "tool_call_id": cid,
})
```

Two implications: (1) on the direct path, the agent's stub-injection already runs after `run_conversation` receives the history, so a 400 from an orphaned call should already be prevented there — which raises the question of whether the WebUI-side strip is even reachable as a fix vs. defense-in-depth; (2) if you do want it WebUI-side, matching the agent's stub-injection keeps the call/result count intact and is lossless, vs. the strip which silently drops context (your own Risks note flags this).

## One thing to verify

The mirrored change in `_api_safe_message_positions` (`api/streaming.py:2746`) now *drops* fully-orphaned no-content assistant entries. That function feeds positional alignment in `_restore_reasoning_metadata` / `_restore_display_reasoning_metadata` (`api/streaming.py:2759,2792`), where `safe_pos` indexes into `updated_messages`. If the agent stub-injected and kept that assistant message, `prev_safe` is now shorter than the post-agent safe set — the `_safe_projection` equality guard prevents data corruption, but reasoning restore could silently no-op for trailing messages. Worth a targeted test asserting reasoning still restores when a prior turn had an orphaned-then-dropped assistant message.

Net: the change is well-formed and tested for the direct path; the gateway-mode framing in the issue may need a gateway/agent-side fix to actually close #3583.

**Comment by bergeouss** [comment]
## Acknowledged — Layering Concern

Thanks for the thorough review. A few notes on the flagged items:

**Gateway path framing**: Agreed — the fix sits on the direct (non-gateway) agent path. The gateway path (`api/gateway_chat.py:308`) only sends prefill + current user message, so orphaned `tool_calls` reconstruction happens gateway-side. The WebUI-side strip is defense-in-depth for the direct path. Issue #3583's title should be updated to reflect both paths.

**Lossy strip vs. agent's stub-injection**: The `_sanitize_messages_for_api` strip runs *before* the agent receives history (it's called at `api/streaming.py:5529` and `api/routes.py:11309`). The agent's stub-injection (`agent_runtime_helpers.py:1842`) runs after `run_conversation` receives the sanitized history. So on the direct path, the WebUI strip is the *first* line of defense — it never reaches the agent. The stub-injection catches any cases that slip through to the agent layer. Both layers are valuable.

**`_api_safe_message_positions` drop + reasoning restore**: Good catch. The `_safe_projection` equality guard at line 2825 will cause a mismatch when a prior assistant message is dropped by pass 3 (since it won't appear in `prev_safe`), causing `safe_pos` and the post-agent position to diverge. This means reasoning restore silently no-ops for all trailing messages after the drop point. The practical impact is limited (the dropped message had no content anyway), but worth documenting. A targeted behavioral test for this edge case would be valuable — we'll add one.

Branch: `bergeouss/hermes-webui:pr/orphaned-tool-calls`

🤖 AI-assisted via Hermes Agent

**Comment by nesquena-hermes** [comment]
Shipped in **v0.51.259** (Release IA) — thank you @rodboev! 🙏 (strip orphaned tool_calls (#3583).) Closing as merged-via-release-stage.
